API for `set`, to replace "array-as-set" operations on arrays

dlongnecke-cray opened the pull request:
This PR implements a first draft proposal of the `set` API. The `set` type is intended to act as a replacement for "array-as-set" operations on arrays, so that such operations may be deprecated. 

For the design of this API, I primarily consulted two existing `set` APIs in other languages:
- Python 3: https://docs.python.org/3/library/stdtypes.html#set
- Ruby 2.6: https://ruby-doc.org/stdlib-2.6.3/libdoc/set/rdoc/Set.html

When critiquing, please consider answers to the following questions:

- Is there a routine you would expect a `set` type to have that is not present?
- Are there any routines in this API that you consider unnecessary, and should be removed?
- Are there any routines you think should be renamed?
- Is the type of a particular routine wrong or unnatural to you?

You may also critique the wording of documentation for routines, though this is not as important as answering the above questions.

---

Link to generated documentation is below:

http://web.us.cray.com/~dlongnecke/set-api/modules/standard/Sets.html

mppf commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,496 @@
+
+      :arg x: The element to add to this set.
+    */
+    proc add(const ref x: eltType) {
```
Wouldn't you want this to have `in` intent just like the list append?

dlongnecke-cray replied:
It is const ref because it makes a copy internally. I can revert to using in intent when #13225 has been resolved.

dlongnecke-cray replied:
Ah wait. I haven't decided on the implementation yet, so `in` intent may well be reasonable here.

mppf replied:
But we're doing API review right now anyway....

dlongnecke-cray replied:
Yes we are, so I can give the method the `in` intent and change it later should it clash with the implementation.

mppf commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,496 @@
+      :arg other: The set to compare against.
+      :return: Whether or not this set and `other` are disjoint.
+    */
+    proc const disjoint(const ref other: set(eltType, ?)): bool {
```
I'd consider naming it `isDisjoint`. I keep trying to interpret `disjoint` as a verb, like "make disjoint"

dlongnecke-cray replied:
I would tend to agree that these are all "is" methods.

mppf commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,496 @@
+      :arg other: The set to compare against.
+      :return: Whether or not this set and `other` intersect.
+    */
+    proc const intersect(const ref other: set(eltType, ?)): bool {
```
I think `intersect` is already a verb and it means to compute the intersection of the set with another set. That would result in a set. However this returns a bool. How about naming it `isIntersecting` ?

mppf commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,496 @@
+      :arg other: The set to compare against.
+      :return: Whether or not this set is a subset of `other`.
+    */
+    proc const subset(const ref other: set(eltType, ?)): bool {
```
Same problem as `intersect`. How about `isSubset` or `containsSet` ?

mppf commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,496 @@
+      :arg other: The set to compare against.
+      :return: Whether or not this set is a proper subset of `other`.
+    */
+    proc const properSubset(const ref other: set(eltType, ?)): bool {
```
`isProperSubset` ?

mppf commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,496 @@
+      :arg other: The set to compare against.
+      :return: Whether or not this set is a superset of `other`.
+    */
+    proc const superset(const ref other: set(eltType, ?)): bool {
```
`isSuperset` ?

mppf commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,496 @@
+      :arg other: The set to compare against.
+      :return: Whether or not this set is a proper superset of `other`.
+    */
+    proc const properSuperset(const ref other: set(eltType, ?)): bool {
```
`isProperSuperset` ?

mppf commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,496 @@
+
+      :throws IllegalArgumentError: If the list contains no such element.
+    */
+    proc remove(const ref x: eltType): eltType throws {
```
Why does this return `eltType` ?

dlongnecke-cray replied:
If `remove` does not return a result, then there is no method capable of removing a element by value and then returning it.

I figure having such a method in the interface would be valuable. However this does conflate the semantics of the `set` remove with the `list` remove.

mppf replied:
OK

mppf commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,496 @@
+    /*
+      Returns a new DefaultRectangular array containing a copy of each of the
+      elements contained in this set. The returned array is not guaranteed
+      to follow any particular ordering.
```
I'd make `toArray` sort the set, personally. Is there a reason you didn't do that? To sort or not could be an option argument.

dlongnecke-cray replied:
I stated "no particular ordering" because an implementation of `set` may not maintain its elements internally in any particular order.

It would be trivial to remove that description and then sort the output before returning it, though.

mppf replied:
Yes, that's what I'm talking about doing

dlongnecke-cray replied:
Ok, that seems reasonable.

mppf commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,496 @@
+    :return: A new set containing the difference between `a` and `b`.
+    :rtype: `set(?t, ?)`
+  */
+  proc setUnion(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t, ?) {
```
I would call this `union`

dlongnecke-cray replied:
I cannot because `union` is a keyword. This is why all of these named methods are prefixed with the word "set", for consistency.

mppf replied:
good point!

dlongnecke-cray replied:
Given that `union` is a keyword, do you have a name that we might use instead?

mppf replied:
`setJoin` or `U` or `or`? I'd probably just let `|` do it though.

mppf commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,496 @@
+    :return: A new set containing the difference between `a` and `b`.
+    :rtype: `set(?t, ?)`
+  */
+  proc |(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t, ?) {
```
For this and the others - I'm not convinced we need both the operator and the free function. It would make slightly more sense if the function were instead a method, so you could write
`a | b` or `a.union(b)`.

dlongnecke-cray replied:
That seems reasonable, but I do think the operators should exist. I will move those to be methods instead.

mppf replied:
How would you feel about only having the operators?

dlongnecke-cray replied:
That would certainly bypass the collision with the `union` keyword. I'm not entirely sure, though. Do we want to establish a precedent for standard library types to favor operator overloads over named functions?

To be sure, if I were writing code using sets I would favor the operators exclusively...

mppf replied:
Well, at this stage, it's reasonable to have some more minimal API and then expand it as experience/user demand requires. Adding functions will be backward-compatible (i.e. not "breaking").

mppf commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,496 @@
+    :return: A new set containing the difference between `a` and `b`.
+    :rtype: `set(?t, ?)`
+  */
+  proc setDifference(const ref a: set(?t, ?),
```
See above

mppf commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,496 @@
+    :return: A new set containing the intersection of `a` and `b`.
+    :rtype: `set(?t, ?)`
+  */
+  proc setIntersection(const ref a: set(?t, ?),
```
See above

mppf commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,496 @@
+    :return: A new set containing the symmetric difference of `a` and `b`.
+    :rtype: `set(?t, ?)`
+  */
+  proc setSymmetricDiff(const ref a: set(?t, ?),
```
See above

mppf commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,496 @@
+    :return: `true` if `a` is a proper subset of `b`.
+    :rtype: `bool`
+  */
+  proc <(const ref a: set(?t, ?), const ref b: set(t, ?)): bool {
```
If we have these comparison routines, why do we need methods like `isProperSubset` ?

dlongnecke-cray commented:
Ok @mppf, is it correct to summarize your suggestions as follows?

- Give the `add` method the `in` intent.
- Eliminate named methods/free functions in favor of operators wherever possible.
- The `toArray` method should sort its output.

mppf commented:
Yes. I think it's OK to have `remove` return something.

dlongnecke-cray commented:
If `remove` can return something, should we change `list` remove to match? Or rename list `remove`? I wonder if you think having consistent semantics matters.

mppf commented:
I think `remove` returning something is relatively harmless in either case. I think it would be fine for `list.remove` to return the removed element.

dlongnecke-cray commented:
Sounds good. I would like that.

dlongnecke-cray commented:
Thanks for the feedback @mppf, I believe I've responded to most of it. If anyone else has feedback on the API I will leave it open till the end of the week.

e-kayrakli commented:
I am thinking whether it would make sense to have a `contains` overload that takes as parameter another set of same type of elements. That overload would simply be a synonym for `isSubset`. Even though I cannot think of an actual use-case right now, it feels somewhat similar to dealing with `1*int` tuples vs `int`s. One potential counter-argument is that it may be a slippery slope. Should we overload `contains` for arrays? Lists? Tuples? Anything iterable?

The similar idea is not applicable to Python because of the relaxed (or lack thereof?) type system. `b in a` for `a` and `b` sets would check whether `b` is a member of `a` as a whole set. I don't know much about Ruby to draw parallels.

cassella commented:
Should there be a `bulkAdd()` sort of interface?  `fromArray()` or `fromIterable()` or an `=` operator whose rhs is any iterable?  An `init=()`?

cassella commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,395 @@
+    /*
+      Iterate over the elements of this set.
+
+      :yields: A reference to one of the elements contained in this set.
```
Other iterator docstrings in `modules/` describe the yields overall, not a single yielded instance.  e.g., 
```
modules/packages/DistributedIters.chpl:  :yields: Indices in the range ``c``.
```

Does it yield them in the same sorted order as `toArray()`?

cassella commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,395 @@
+      elements contained in this set. The returned array is sorted according
+      to the default sort order of its elements.
+    */
+    proc const toArray(): [] eltType {
```
I'd make the sorting optional, or make the order unspecified and let the caller sort the resulting array as desired.  Or allow the user to provide their own `comparator`.  Unless you're expecting to be able to provide that particular sorted order efficiently?

mppf replied:
Good point about the comparator. Also it's important to note that not all types are currently sortable.

One reasonable strategy would be to allow a comparator argument, e.g. with `proc const toArray(comparator:? = none)`. The body of this function can check if comparator is not `none` and sort in that event. I like this adjustment to the strategy because:
 * it makes a common use case (wanting to sort when creating an array) easy
 * it still allows for flexibility

I am not expecting that the sort is any more efficient than if the user sorted the result of toArray themselves.

cassella commented on modules/standard/Sets.chpl:
```diff
@@ -0,0 +1,395 @@
+    :arg a: A set to take the difference of.
+    :arg b: A set to take the difference of.
+
+    :return: A new set containing the difference between `a` and `b`.
```
That's a funny thing for `proc |()` to return, and for its args to be for.

dlongnecke-cray replied:
It seems to me that `|` is the natural operator to overload with the semantics of "union", do you have an alternative in mind?

mppf replied:
I think @cassella was just pointing out that the chpldoc string uses "difference" in several places in the arg/return descriptions

dlongnecke-cray replied:
Ha ha!

Thanks for the catch @cassella, I'll be honest I definitely copy/pasted the description from the difference docstring and forgot to change it.

mppf commented:
> I am thinking whether it would make sense to have a contains overload that takes as parameter another set of same type of elements. That overload would simply be a synonym for isSubset. 

When there are sets containing sets, the compiler might get confused about which `contains` to call.

However I really like the idea of moving to more consistent terminology - ie. if we already have `contains`, can we have `containsSet` instead of `isSubset` ?

dlongnecke-cray commented:
In general, I think that all containers in the standard library should have some sort of standardized "bulkAdd" method that is generic over any iterable, but I'm not quite sure what form that will take yet (and I would like to go back and edit the `list` API to account for that).

To me, sorting the output of `toArray()` feels rather strange, especially when the `list` variant does not have such behavior. 

I myself would prefer making the sort explicit (IE, the user calls sort over the newly created array) over overloading the `toArray()` method with a comparator.

However, any changes made in the semantics of `toArray()` for set should preferably be mirrored in `list` as well.

@cassella Semantically, I've only ever envisioned contains as checking for the existence of a single element, not as a "subset/superset" style operator.

Classically, most languages I've seen (at least, Python and Ruby) tend to use `|` as the "union" operator. Were you envisioning another operator (say, `+`) be used here instead?

To @mppf, @bradcray thought it would be good to define the op-equals (`|=`, `&=`, `-=`, `^=`) for the `set` type. Do you think they should be member methods or free functions?

bradcray commented:
> To @mppf, @bradcray thought it would be good to define the op-equals (|=, &=, -=, ^=) for the set type. Do you think they should be member methods or free functions?

Chapel doesn't really support operators-as-methods at present, so these need to be free functions (as they are on associative domains today).

bradcray commented:
> Were you envisioning another operator (say, +) be used here instead?

I think currently, associative and sparse domains support both `+` and `|` for this (i.e., they're synonyms for one another, for better or worse)

mppf commented:
> I myself would prefer making the sort explicit (IE, the user calls sort over the newly created array) over overloading the toArray() method with a comparator.

I'm fine with leaving the sort out of toArray.

dlongnecke-cray commented:
What should the name of this `bulkAdd` method be? The `list` type currently has a method called `extend`, which appends elements to the back of the list. 

Would calling the `bulkAdd` method `extend`, work? Or should a different name be used instead?

dlongnecke-cray commented:
I've added `op-equals` routines for the union, intersection, difference, and symmetric difference operators respectively.

I've also added a new `addBulk` routine that is generic over any iterable.

If others are ok with the convention of appending "Bulk" to any operation which can accept an iterator, then I am considering the following changes to `list`:

- Rename `extend` to `appendBulk`
- Add a new method called `insertBulk`

e-kayrakli commented:
I am against `addBulk`, because we already have a `bulkAdd` for sparse domains, and it would be too confusing. See: https://chapel-lang.org/docs/builtins/ChapelArray.html#ChapelArray.bulkAdd

That doesn't mean that `bulkAdd` is better than `addBulk`, but I am just all for consistency.

e-kayrakli commented:
Re `containsSet` I would think that the compile would ideally be able to resolve that (if this is a set of sets of ints, and the arg is a set of ints, then it is a member check, if the arg is a set of set of ints, it is a subset check) but maybe I am underestimating the difficulty of handling recursive generics there as you go deeper. I can live without `containsSet` but have no problem against it.

dlongnecke-cray commented:
Ok, that seems fair @e-kayrakli, I can rename the routine to `addBulk` for consistency.

As for `contains`...To me the semantics of contains seem to be clearly defined as a member check. If you want to see if another set is contained in a given set, why not use the subset operator? Overloading `contains` in the fashion you are described seems to me to be cramming too much semantics into one function, while something like `containsSet` would be redundant (I do see that you're fine with either).

dlongnecke-cray commented:
After some internal debate, I have removed the `bulkAdd` routine for now. I have however changed the overload of `init=` to be generic over any iterable rather than expect a `set` type. This is a change that I can see carrying over to the `init=` for the `list` type as well.

dlongnecke-cray commented:
This PR has produced some very useful discussion! I will be closing this rather than merging the yet unimplemented API. The fully implemented set type will be up for review sometime in the next week.
